Faster transitive closure computation for DAGs

alabarre opened the pull request:
This is an improved algorithm for computing the transitive closure of a DAG using topological_sort (so an exception will be raised if the input graph is not a DAG): traverse vertices following a reverse topological order, and connect each vertex to its descendants at distance 2 as we go. The topological order guarantees that each successor v of a node u is connected to all descendants of u in the transitive closure, and therefore it is enough to connect u to nodes at distance 2 from u rather than attempting to explore all descendants.

The algorithm is probably folklore but I didn't find a reference in my books, feel free to add one if you can.

I also added a function to compute the descendants of a node at a fixed distance. `bfs_successors` worked just as well in my tests but resulted in code that was harder to read. Moreover, the new function can be of general interest and is less memory-hungry since each iteration only stores nodes at a fixed distance, whereas `bfs_successors` will keep previous nodes as well.

A taste of the kind of results one can expect using the time function (I used assert to check that both computed closures have the same vertex and edge sets): 

    $ python3 better_transitive_closure_dag.py 
    Testing transitive closure on a random DAG with 800 nodes and 239532 edges
    Running time for NetworkX's TC:             17.484230041503906
    Running time for TC dag-optimized:           3.222315549850464

I'd gladly add my benchmarks but I'm at loss as to where to include them. I shamelessly copied the tests for `transitive_closure` in the meantime.

dschult commented:
Questions about ```descendants_at_distance```:

- It looks like you don't need a deque. I mean that the order of of the nodes is not important so appendleft could just be append with a list or even add with a set.  Is there any other reason to use deque that I am missing? If not, then we should use list or set.
- As noted in and near your TODO, the could be changes to use ```G[vertex]``` instead of ```G.successors(vertex)``` and it would work for directed or undirected graphs. Then it should be put into breadth_first_search.py since it is quite general and perhaps helpful to other routines.

```transitive_closure``` does feel like a misfit in dag.py.  Can you recommend another place for it?

In ```transitive_closure_dag``` you use a list topo_order so you can index it. But it is perhaps better to use ```reversed(topo_order)``` instead of the indexing. That allows you to not require a list, only an iterable. 

The first line of the doc_string should be split from the rest of that first paragraph.  Sphinx uses the first line, followed by a blank line, as a short summary when it creates links to the function in the reference.  So:  one-line, blank line, paragraphs separated by blank lines.

This looks quite good. Thank you.

alabarre commented:
> It looks like you don't need a deque. I mean that the order of of the nodes is not important so appendleft could just be append with a list or even add with a set.  Is there any other reason to use deque that I am missing? If not, then we should use list or set.


Indeed, I was too focused on the breadth-first search aspect. I've replaced the deque with a set.

> As noted in and near your TODO, the could be changes to use `G[vertex]` instead of `G.successors(vertex)` and it would work for directed or undirected graphs. Then it should be put into breadth_first_search.py since it is quite general and perhaps helpful to other routines.

Fixed, thanks. I also removed the "Notes" section which thereby became obsolete.

> `transitive_closure` does feel like a misfit in dag.py. Can you recommend another place for it?

I see two options:

1. move it to a placeholder for directed graph algorithms; there doesn't seem to be one yet. 
1. move it to a dedicated file for transitive closure algorithms, along with the one for DAG's, which would then also include algorithms for undirected graphs (although in that case I can't think of anything other than replacing every connected component with the complete subgraph on the corresponding vertex set). I don't know if there exist enough improved algorithms for specialised cases to warrant this however.

> In `transitive_closure_dag` you use a list topo_order so you can index it. But it is perhaps better to use `reversed(topo_order)` instead of the indexing. That allows you to not require a list, only an iterable.

I replaced the slice with a call to `reversed`, but I still need to cast `topo_order` as a `list` when invoking `topological_sort`, since it is a generator and `reversed` demands a sequence as input.

> The first line of the doc_string should be split from the rest of that first paragraph. Sphinx uses the first line, followed by a blank line, as a short summary when it creates links to the function in the reference. So: one-line, blank line, paragraphs separated by blank lines.

I didn't know that, I fixed it.

dschult commented:
Hmmm.. not sure the reversed vs indexing suggestion helped then... and I learned something. :)

Let's leave transitive closure where it is for now. If we get anything else related to it I'd want to refactor into a separate module. 

Can you do one more check using pycodestyle to check for PEP8 compliance. Is there anything else for this PR?

alabarre commented:
> Hmmm.. not sure the reversed vs indexing suggestion helped then... and I learned something. :)

Well, there are discussions about which approach is faster ... in any case, it is probably more readable as it now stands, and your argument made sense. Let's leave it as is until someone else has a good reason to change it again.

> Can you do one more check using pycodestyle to check for PEP8 compliance. Is there anything else for this PR?

Done, I fixed a minor issue (line too long for one import). There are two other PEP8 issues in `is_aperiodic` (a poorly named variable `l`, and also `gcd` is apparently deprecated) but since I'm not the author I'd rather leave it to him / her to fix it.

I have nothing more to add or change so we can probably close the discussion.
